Add systemd-257 package to the core-kit

vigh-m opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Related #536

**Description of changes:**

Add a functioning build of the latest stable systemd-257 to the core-kit. This change does not build `systemd-cryptsetup` due to incompatibilities between `aws-lc` and `openssl` support.

The initial build resulted in a version of systemd that build but did not result in successful launches of BR AMIs. The subsequent changes are all to get the AMIs to a working stage

**Testing done:**
Tested the following variants built with systemd-257 and running a sample workload.

Variants Tested:
- Long term soak testing on aws-k8s-1.32 and aws-k8s-1.33 varaints
- [x] aws-k8s-1.28
- [x] aws-k8s-1.29
- [x] aws-k8s-1.30
- [x] aws-k8s-1.31
- [x] aws-k8s-1.32
- [x] aws-k8s-1.33
- [x] aws-k8s-1.28-nvidia
- [x] aws-k8s-1.29-nvidia
- [x] aws-k8s-1.30-nvidia
- [x] aws-k8s-1.31-nvidia
- [x] aws-k8s-1.32-nvidia
- [x] aws-k8s-1.33-nvidia
- [x] aws-ecs-2
- [x] aws-ecs-2-nvidia

## Systemd v257 to v252 NEWS Summary for Bottlerocket

This summary only includes what I deemed impactful to Bottlerocket. Refer to the full [NEWS] for greater detail.

[NEWS]: https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS

### Major changes

* [Support for cgroup v1 ('legacy' and 'hybrid' hierarchies) is now considered obsolete](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L54) and systemd by default will ignore configuration that enables them. To forcibly reenable cgroup v1 support, `SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1` must additionally be set on the kernel command line.

### Compatibility Breaks

* [Support for `split-usr` and `unmerged-usr` has been removed](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L1944). This will not affect Bottlerocket since `-Dsplit-usr` was already `false`.

### New features to explore

* [A new "soft-reboot" mechanism has been added to the service manager](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L2741). A "soft reboot" is similar to a regular reboot, except that it affects userspace only: the service manager shuts down any running services and other units, then optionally switches into a new root file system (mounted to `/run/nextroot/`)

* [The `RestartMode=` setting gained a new "debug" value](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L197). 
    If specified and the service fails so that it shall be restarted it is invoked in "debugging mode" which will (among other things) set `LogLevel` to `debug` temporarily

* [A new service type `Type=notify-reload` is defined](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L1233). When such a unit is reloaded a UNIX process signal (typically SIGHUP) is sent to the main service process. The manager will then wait until it receives a "RELOADING=1" followed by a "READY=1" notification from the unit as response (via sd_notify()).

### Changes in systemd itself

* [Services are now started via the systemd-executor](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L1994C9-L2012C51)
    systemd-executor, receives the configuration to apply via memfd, and sets up the process before exec'ing the target executable. 

### Changes in systemd-resolved
* [systemd-resolved's configuration may now be reloaded without restarting the service](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L1322)

### Changes in systemd-networkd

* [systemd-networkd now provides a basic Varlink interface](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L1233)

###  [Unprivileged User Namespaces & Mounts](https://github.com/systemd/systemd/blob/1d51b4d092a67812ec3433c93ba2d3a66a16a6fb/NEWS#L1788):

* A small new service systemd-nsresourced.service has been added.
    It provides a Varlink IPC API that assigns a free, transiently allocated 64K UID/GID range to an uninitialized user namespace a client provides. It may be used to implement unprivileged container managers
          and other programs that need dynamic user ID ranges

* A small new service systemd-mountfsd.service has been added. 
It provides a Varlink IPC API for mounting DDI images, and returning a set  of mount file descriptors for it.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

vigh-m committed:
systemd: initial build of v257

Systemd-252 patches 1001, 1002, and 1003 are no longer needed as they are
included upstream.

Patches 9001, 9007, 9014, 9016, and 9018 are removed since the changes
are no longer relevant

Patches 9002, 9003, 9005, 9006, 9010, 9011, 9012, 9013, 9015, and 9016
were reworked for this update

Patches 9004 and 9009 have been reworked as drop-ins

vigh-m committed:
selinux: add rules for systemd-executor process

Upstream systemd introduced systemd-executor as a new component that
handles process execution for systemd services. This executor requires
SELinux permissions to function correctly within Bottlerocket's
security policy.

Reference: https://github.com/systemd/systemd/blob/main/NEWS#L2092-L2110

vigh-m committed:
systemd-resolved: disable PrivateTmp feature

Upstream systemd changed the default behavior to enable PrivateTmp by
default for systemd-resolved. The PrivateTmp feature isolates temporary
files by creating a private mount namespace for /tmp, which conflicts
with Bottlerocket's current filesystem layout.

Reference: https://www.man7.org/linux//man-pages/man5/systemd.exec.5.html

arnaldo2792 commented on packages/systemd-257/9017-prevent-bootctl-build.patch:
```diff
@@ -0,0 +1,29 @@
+From 646e0f569076f2a68ed2c17c215d464db82aa4a7 Mon Sep 17 00:00:00 2001
```
Is this a revert of a commit or a manual patch? If a manual patch, can you document why the manual revert didn't work?

bcressey commented on packages/systemd-257/9001-use-absolute-path-for-var-run-symlink.patch:
```diff
@@ -0,0 +1,29 @@
+From 2e4d5ecb4c47a9d2d20cb4de42484eda032a5262 Mon Sep 17 00:00:00 2001
+From: Ben Cressey <bcressey@amazon.com>
+Date: Mon, 3 Jan 2022 20:22:17 +0000
+Subject: [PATCH] use absolute path for /var/run symlink
```
We should be able to drop this patch because `mask-local-var.service` makes it so there's no longer a visible `/local/var` path on the host.

bcressey commented on packages/systemd-257/9003-machine-id-setup-generate-stable-ID-under-Xen-and-VM.patch:
```diff
@@ -0,0 +1,31 @@
+From dd1d0221bb0ff143277faa4a7341e290a3941587 Mon Sep 17 00:00:00 2001
+From: Ben Cressey <bcressey@amazon.com>
+Date: Mon, 3 Jan 2022 21:57:11 +0000
+Subject: [PATCH] machine-id-setup: generate stable ID under Xen and VMware
```
nit: it's only for `VIRTUALIZATION_VMWARE` at this point.

bcressey commented on packages/systemd-257/9009-units-keep-modprobe-service-units-running.patch:
```diff
@@ -0,0 +1,24 @@
+ [Service]
+ Type=oneshot
+ ExecStart=-/sbin/modprobe -abq %i
++RemainAfterExit=true
```
nit: this could be a drop-in rather than a patch

bcressey commented on packages/systemd-257/9004-units-mount-tmp-with-noexec.patch:
```diff
@@ -0,0 +1,23 @@
+ Where=/tmp
+ Type=tmpfs
+-Options=mode=1777,strictatime,nosuid,nodev,size=50%%,nr_inodes=1m
++Options=mode=1777,strictatime,nosuid,nodev,noexec,size=50%%,nr_inodes=1m
```
This could be a drop-in instead of a patch

bcressey commented on packages/systemd-257/9013-meson-make-gpt-auto-generator-selectable-at-build-ti.patch:
```diff
@@ -0,0 +1,53 @@
+Date: Thu, 9 Nov 2023 15:02:14 +0000
+Subject: [PATCH] meson: make gpt-auto-generator selectable at build-time
+
+Signed-off-by: Markus Boehme <markubo@amazon.com>
```
For patches that change significantly, you should add your own notes:
```
[vighmah: short summary of changes]
Signed-off-by: ...
```

Or:
```
[vighmah:
  - note 1
  - note 2
]
Signed-off-by: ...
```

bcressey commented on packages/release/systemd-resolved-service-private-tmp.conf:
```diff
@@ -0,0 +1,2 @@
+[Service]
+PrivateTmp=false
```
What is the nature of the conflict with Bottlerocket? Can you add a brief comment here?

bcressey commented on packages/systemd-257/9018-Disable-persistent-storage-notification-for-systemd-networkd.patch:
```diff
@@ -0,0 +1,24 @@
+ After=systemd-networkd.socket systemd-udevd.service network-pre.target systemd-sysusers.service systemd-sysctl.service
+ Before=network.target multi-user.target shutdown.target initrd-switch-root.target
+ Conflicts=shutdown.target initrd-switch-root.target
+-Wants=systemd-networkd.socket network.target systemd-networkd-persistent-storage.service
```
Can we just exclude `systemd-networkd-persistent-storage.service` if we don't want the notification? It's OK to "want" a missing unit.

bcressey commented on packages/systemd-257/journald.conf:
```diff
@@ -0,0 +1,3 @@
+[Journal]
+Storage=persistent
+SplitMode=none
```
nit: we could move some of these files to `release` so they're not duplicated across systemd packages:
* issue
* journald.conf
* systemd-journald.conf
* systemd-modules-load.conf
* systemd-sysusers.conf

bcressey commented on packages/systemd-257/9011-core-mount-increase-mount-rate-limit-burst-to-25.patch:
```diff
@@ -0,0 +1,29 @@
+         if (!m->mount_monitor) {
+                 usec_t mount_rate_limit_interval = 1 * USEC_PER_SEC;
+-                unsigned mount_rate_limit_burst = 5;
++                unsigned mount_rate_limit_burst = 25;
```
It'd be good to validate that this limit is still high enough, by adding a warning log if we hit the rate limit. We might have to increase it again now that the boot is faster.

bcressey reviewed (commented):
Reviewed everything but the spec.

vigh-m committed:
release: move common files from systemd-* to release

There are common files between systemd-252 and systemd-257 which
would be better situated in the `release` package

mikn commented:
This is great! Is there a timeline to re-enable systemd-cryptsetup? I just hit an issue with TPM initialization - we initially did it manually, now we tried changing over to using systemd-tpm2-setup (introduced in v255), but this initializes the TPM in a way that is incompatible with v252. Would be great to be able to switch to v257 on the Bottlerocket side so we could remove a lot of our custom TPM dance.

bcressey commented on packages/systemd-257/tmp-mount.conf:
```diff
@@ -0,0 +1,2 @@
+[Service]
+Options=mode=noexec
```
Does this preserve all the other default options?
```
Options=mode=1777,strictatime,nosuid,nodev,size=50%%,nr_inodes=1m
```

vigh-m replied:
Good catch. It did not. I'm reworking it to include default options. I assumed it would behave like other drop-ins and merge with the unit files. 

---

With just `noexec`

```bash
bash-5.1# systemctl show tmp.mount | grep Options
Options=rw,noexec,relatime,seclabel
SloppyOptions=no

```

With all the options from upstream + `noexec`

```bash
bash-5.1# systemctl show tmp.mount | grep Options
Options=rw,nosuid,nodev,noexec,seclabel,nr_inodes=1048576
SloppyOptions=no
bash-5.1# systemctl status tmp.mount
● tmp.mount - Temporary Directory /tmp
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/tmp.mount; static)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/tmp.mount.d
             └─10-no-exec.conf
     Active: active (mounted) since Thu 2025-08-14 22:15:25 UTC; 1min 5s ago
 Invocation: b0004224a3b64916b0403c4c229ad993
      Where: /tmp
       What: tmpfs
       Docs: https://systemd.io/TEMPORARY_DIRECTORIES
             man:file-hierarchy(7)
             https://systemd.io/API_FILE_SYSTEMS
      Tasks: 0 (limit: 9313)
     Memory: 12K (peak: 1.1M)
        CPU: 5ms
     CGroup: /system.slice/tmp.mount

Notice: journal has been rotated since unit was started, output may be incomplete.

```

bcressey commented on packages/systemd-257/systemd-modprobe.conf:
```diff
@@ -0,0 +1,2 @@
+[Service]
+RemainAfterExit=true
```
This drop-in and the other new one might as well go in `release` also. They're compatible with 252 and that way we won't have to move them when we add the next version of systemd.

bcressey commented on packages/systemd-257/9001-core-add-separate-timeout-for-system-shutdown.patch:
```diff
@@ -0,0 +1,84 @@
+Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>
+
+[vighmah:
+Ported to work with the changes introduced in 
```
nit: trailing space
```suggestion
Ported to work with the changes introduced in
```

bcressey commented on packages/systemd-257/9002-machine-id-setup-generate-stable-ID-under-VM.patch:
```diff
@@ -0,0 +1,32 @@
+Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>
+
+[vighmah:
+Ported to work with the changes introduced in 
```
nit:
```suggestion
Ported to work with the changes introduced in
```

bcressey commented on packages/systemd-257/9001-core-add-separate-timeout-for-system-shutdown.patch:
```diff
@@ -0,0 +1,84 @@
+From e66920d0a95a3e53b24a2f270420a85d0f2395ea Mon Sep 17 00:00:00 2001
+From: Ben Cressey <bcressey@amazon.com>
+Date: Tue, 10 Mar 2020 20:30:10 +0000
+Subject: [PATCH] core: add separate timeout for system shutdown
```
I'm somewhat inclined to drop this patch; I don't think we should reach either the patched 10 second or the default 30 second timeout in production scenarios where workloads have been drained prior to a reboot attempt.

It'd be good to reboot a node without this patch and see if we end up stuck for 30 seconds anywhere.

bcressey commented on packages/systemd-257/9003-mount-setup-apply-noexec-to-more-mounts.patch:
```diff
@@ -0,0 +1,39 @@
+-        { "tmpfs",       "/run",                      "tmpfs",      "mode=0755" TMPFS_LIMITS_RUN,               MS_NOSUID|MS_NODEV|MS_STRICTATIME,
++        { "tmpfs",       "/run",                      "tmpfs",      "mode=0755" TMPFS_LIMITS_RUN,               MS_NOSUID|MS_NOEXEC|MS_NODEV|MS_STRICTATIME,
+           NULL,          MNT_FATAL|MNT_IN_CONTAINER },
+         { "cgroup2",     "/sys/fs/cgroup",            "cgroup2",    "nsdelegate,memory_recursiveprot",          MS_NOSUID|MS_NOEXEC|MS_NODEV,
```
This is where the hybrid or unified cgroup mounts used to be set up - it'd be worth double-checking that the system ends up with a `tmpfs` mount on `/sys/fs/cgroup` as expected with the right kernel command-line overrides.

vigh-m replied:
Tested it out. The variant kernel params need the following:
```toml
  kernel-parameters = [
    "console=tty0",
    "console=ttyS0,115200n8",
    "net.ifnames=0",
    "netdog.default-interface=eth0:dhcp4,dhcp6?",
    "quiet",
+   "SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1",
+   "systemd.unified_cgroup_hierarchy=0",
  ]
```
Which is what upstream [calls out](https://github.com/systemd/systemd/blob/5e38d199a623563698ab4a69acbbe3afa9135198/src/shared/cgroup-setup.c#L167)

After that I see the right tmpfs mounts for /sys/fs/cgroup
```bash
bash-5.1# findmnt /sys/fs/cgroup/
TARGET         SOURCE FSTYPE OPTIONS
/sys/fs/cgroup tmpfs  tmpfs  ro,nosuid,nodev,noexec,seclabel,size=4096k,nr_inodes=1024,mode=755
```

I also booted up a Bottlerocket k8s-1.25 AMI and verified that it's the same

bcressey commented on packages/systemd-257/9003-units-mount-tmp-with-noexec.patch:
```diff
@@ -0,0 +1,23 @@
+From d9eb89767267d43f13c30ce74f6e9c34c4000274 Mon Sep 17 00:00:00 2001
+From: Ben Cressey <bcressey@amazon.com>
+Date: Sat, 28 May 2022 03:52:28 +0000
+Subject: [PATCH] units: mount /tmp with noexec
```
Do we still need this patch since we have the `tmp-mount.conf` drop in?

bcressey commented on packages/systemd-257/9007-systemd-networkd-Conditionalize-hostnamed-timezoned-.patch:
```diff
@@ -0,0 +1,125 @@
+From 1539259546fe7f686d07e256d700ee2929b78064 Mon Sep 17 00:00:00 2001
```
Can you confirm that the generated `systemd.pc` pkgconfig file has the correct sysroot prefix, and not just `/usr`? That was the old 9007 patch.

vigh-m replied:
Yup. Cracked open the `-devel` rpm that's generated and I can see the following
```
bash-5.2$ cat systemd.pc | grep prefix
root_prefix=/x86_64-bottlerocket-linux-gnu/sys-root/usr
rootprefix=${root_prefix}
prefix=${rootprefix}
bash-5.2$ cat libsystemd.pc | grep prefix
prefix=/x86_64-bottlerocket-linux-gnu/sys-root/usr
exec_prefix=/x86_64-bottlerocket-linux-gnu/sys-root/usr
```

bcressey commented on packages/systemd-257/9010-meson-make-gpt-auto-generator-selectable-at-build-ti.patch:
```diff
@@ -0,0 +1,60 @@
+From 33a689a000d56444ed47803b2fbec4376814e633 Mon Sep 17 00:00:00 2001
+From: Markus Boehme <markubo@amazon.com>
+Date: Thu, 9 Nov 2023 15:02:14 +0000
+Subject: [PATCH] meson: make gpt-auto-generator selectable at build-time
```
Now that we have an explicit file list for all packaged files, it might be easier to drop this patch and simply exclude the `systemd-gpt-auto-generator` files.

vigh-m replied:
Yep! I was able to remove the patch with minimal fuss

bcressey commented on packages/systemd-257/9014-prevent-bootctl-build.patch:
```diff
@@ -0,0 +1,29 @@
+
+Re-adding the condition to prevent bootctl build removed in
+https://github.com/systemd/systemd/pull/29708 which ends up causing
+bootctl to build as a default.
```
Can we let `bootctl` build and just exclude it? Previously it depended on `gnu-efi` which was a lot to package.

bcressey commented on packages/release/systemd-resolved-service-private-tmp.conf:
```diff
@@ -0,0 +1,8 @@
+# our read-only filesystem.
+# Setting to false has the added benefit of separating the systemd-resolved from the
+# setting up of tmpfiles and the local-fs.target
+# Upstream Issue: https://github.com/systemd/systemd/issues/37096
```
nit: let's drop this comment since it's spamming the upstream issue with every version of this commit

Have you given the maintainer's suggestions a try?
1. add `/var/tmp` as another directory in `filesystem.spec` so the mount point always 
2. or add `TemporaryFileSystem=/var` to the unit instead of this drop-in
3. or set `PrivateTmp=disconnected` to avoid any dependency on the host's directories

vigh-m replied:
I've updated the comment. But essentially, `disconnected` still causes an error. From [systemd-exec man](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html):
```
If "disconnected", the directories will be backed by a completely new tmpfs instance, meaning that the storage is fully disconnected from the host namespace.
```

Ideally, we want it to be `false` since we are working towards decoupling the filesystem and network and this will get us there.

bcressey commented on packages/release/release.spec:
```diff
@@ -100,6 +100,14 @@ Source1102: systemd-networkd-service-env.conf
 Source1104: aws-config.conf
 Source1105: wait-for-selinux-policy.conf
+Source1106: systemd-resolved-service-private-tmp.conf
+Source1107: systemd-tmpfiles.conf
```
This file should be `release-systemd-tmpfiles.conf`, and this file should go in the same range as the other tmpfiles snippets.

bcressey commented on packages/release/release.spec:
```diff
@@ -100,6 +100,14 @@ Source1102: systemd-networkd-service-env.conf
+Source1109: journald.conf
+Source1110: issue
+Source1111: systemd-journald.conf
+Source1112: systemd-sysusers.conf
```
You've added all these files to the "Drop-in units to override defaults" block of source files, but most of them are not actually unit drop-ins. Please move them to the correct sections.

For naming, the convention is basically: if it's a drop-in config file, it's prefixed with the owning package's name. If it's a unit drop-in, it's prefixed with the actual unit name, and has a short token in the filename to indicate what the drop-in does.

This distinction was previously obscured because the package's name was `systemd` and the affected units were also named `systemd-*`.

1. `systemd-tmpfiles.conf` -> merge into `release-tmpfiles.conf`, and also merge `release-snapshotter-tmpfiles.conf` while you're at it.
2. `systemd-modules-load.conf` -> `release-modules-load.conf`
3. `journald.conf` -> `release-systemd-journald.conf`
4. `issue` is OK
5. `systemd-journald.conf` -> `systemd-journald-compat.conf`
6. `systemd-sysusers.conf` -> `systemd-sysusers-selinux.conf`

bcressey commented on packages/release/release.spec:
```diff
@@ -368,6 +399,12 @@ ln -s preconfigured.target %{buildroot}%{_cross_unitdir}/default.target
+%{_cross_libdir}/modules-load.d/nf_conntrack.conf
+%{_cross_unitdir}/systemd-sysusers.service.d/systemd-sysusers.conf
+%{_cross_unitdir}/systemd-journald.service.d/systemd-journald.conf
+%{_cross_libdir}/systemd/journald.conf.d/journald.conf
```
These entries should be moved to be adjacent to other files in or around the same directories, rather than appended to the end of the list.

bcressey commented on packages/systemd-252/systemd-252.spec:
```diff
@@ -700,7 +667,7 @@ find %{buildroot} -type f -name README -print -delete
 
 %dir %{_cross_factorydir}
-%{_cross_factorydir}%{_cross_sysconfdir}/issue
+%exclude %{_cross_factorydir}%{_cross_sysconfdir}/issue
```
I'm surprised this would be required - is there a default `issue` file that's added to `/usr/share/factory/etc`?

vigh-m replied:
Yeah. [This](https://github.com/systemd/systemd/blob/4fd9b83c209b34ada9ba7111123a6eb7acc54868/factory/etc/issue) gets installed and so I excluded it

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+# Skip check-rpaths since we expect them for systemd.
+%global __brp_check_rpaths %{nil}
+
+%global debug_package %{nil}
```
This is almost always an error and you should fix whatever is causing the debuginfo package not to build correctly.
```suggestion
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+# way to be configured at build time or during execution first.
+Patch9001: 9001-core-add-separate-timeout-for-system-shutdown.patch
+
+# TODO: this could be submitted upstream as well, but needs to account for
+# the dom0 case first, where the UUID is all zeroes and hence not unique.
+Patch9002: 9002-machine-id-setup-generate-stable-ID-under-VM.patch
```
Comment no longer applies now that it only affects VMware machine IDs.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+ -Dremote=disabled
+ -Dnss-myhostname=false
+ -Dnss-mymachines=disabled
+ -Dnss-resolve=enabled
```
We don't use the NSS resolver plugin now - it caused a behavior change with Go DNS resolution - so this can likely be disabled.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+ -Dpasswdqc=disabled
+ -Dpwquality=disabled
+ -Dmicrohttpd=disabled
+ -Dlibcryptsetup=disabled
+ -Dlibcryptsetup-plugins=disabled
```
Can you move all the cryptsetup and TPM options to their own block, so it's clear what will need to be flipped to turn it back on?

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+
+%install
+%cross_meson_install
+
+
```
nit: remove these added newlines
```suggestion
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%{_cross_sbindir}/shutdown
+
+%exclude %{_cross_bindir}/systemd-vpick
+%exclude %{_cross_sbindir}/mount.ddi
```
This is just a symlink to `systemd-dissect`, so it's fine to keep it.
```suggestion
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%{_cross_sbindir}/reboot
+%{_cross_sbindir}/shutdown
+
+%exclude %{_cross_bindir}/systemd-vpick
```
I could go either way here, but seems OK to keep this utility:
```suggestion
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%{_cross_libdir}/systemd/systemd-sysctl
+%{_cross_libdir}/systemd/systemd-sysroot-fstab-check
+%{_cross_libdir}/systemd/systemd-udevd
+%{_cross_libdir}/systemd/profile.d/70-systemd-shell-extra.sh
```
exclude since it's a shell script:
```suggestion
%exclude %{_cross_libdir}/systemd/profile.d/70-systemd-shell-extra.sh
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%{_cross_datadir}/dbus-1/system.d/org.freedesktop.login1.conf
+%{_cross_datadir}/dbus-1/system.d/org.freedesktop.systemd1.conf
+%exclude %{_cross_datadir}/dbus-1/system-services
+%exclude %{_cross_datadir}/factory/etc/vconsole.conf
```
`vconsole.conf` should go with the other `%{_cross_factorydir}%{_cross_sysconfdir}` excludes.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%{_cross_datadir}/dbus-1/system.d/org.freedesktop.systemd1.conf
+%exclude %{_cross_datadir}/dbus-1/system-services
+%exclude %{_cross_datadir}/factory/etc/vconsole.conf
+%exclude %{_cross_datadir}/mime/packages/io.systemd.xml
```
This should go at the end of the main package's `%exclude` list with other miscellaneous desktop stuff.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%{_cross_udevrulesdir}/80-net-setup-link.rules
+%{_cross_udevrulesdir}/81-net-dhcp.rules
+%{_cross_udevrulesdir}/99-systemd.rules
+
```
remove extra newline
```suggestion
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+# reboot but avoids setting up a storage dependency for systemd-networkd
+%exclude %{_cross_libdir}/systemd/system/systemd-networkd-persistent-storage.service
+
+
+%exclude %{_cross_libdir}/udev/iocost
+%exclude %{_cross_libdir}/udev/rules.d/60-dmi-id.rules
+%exclude %{_cross_libdir}/udev/rules.d/60-persistent-storage-mtd.rules
+%exclude %{_cross_libdir}/udev/rules.d/90-iocost.rules
```
Let's keep all of these, and please fold them into the existing `%dir %{_cross_libdir}/udev` and `%dir %{_cross_udevrulesdir}` sections.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%exclude %{_cross_libdir}/udev/rules.d/90-iocost.rules
+
+%exclude %{_cross_libdir}/tmpfiles.d/20-systemd-stub.conf
+%exclude %{_cross_libdir}/tmpfiles.d/credstore.conf
```
We can probably keep `credstore.conf`.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+
+%exclude %{_cross_libdir}/systemd/repart/*
+%exclude %{_cross_libdir}/systemd/system-generators/systemd-ssh-generator
+%exclude %{_cross_libdir}/systemd/system-generators/systemd-tpm2-generator
+
+%exclude %{_cross_libdir}/systemd/system/tpm2.target
```
These two seem useful.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%exclude %{_cross_libdir}/systemd/systemd-battery-check
+%exclude %{_cross_libdir}/systemd/systemd-ssh-proxy
+
+%exclude %{_cross_libdir}/systemd/system/capsule.slice
+%exclude %{_cross_libdir}/systemd/system/capsule@.service
+%exclude %{_cross_libdir}/systemd/system/sockets.target.wants/systemd-creds.socket
+%exclude %{_cross_libdir}/systemd/system/soft-reboot.target
+%exclude %{_cross_libdir}/systemd/system/ssh-access.target
+%exclude %{_cross_libdir}/systemd/system/sysinit.target.wants/systemd-tmpfiles-setup-dev-early.service
+%exclude %{_cross_libdir}/systemd/system/systemd-creds.socket
+%exclude %{_cross_libdir}/systemd/system/systemd-creds@.service
+%exclude %{_cross_libdir}/systemd/system/systemd-growfs-root.service
+%exclude %{_cross_libdir}/systemd/system/systemd-growfs@.service
+%exclude %{_cross_libdir}/systemd/system/systemd-journald-sync@.service
+%exclude %{_cross_libdir}/systemd/system/systemd-soft-reboot.service
+%exclude %{_cross_libdir}/systemd/system/systemd-tmpfiles-setup-dev-early.service
+%exclude %{_cross_libdir}/systemd/system/systemd-udev-load-credentials.service
```
Unit-related includes and excludes should be in the `%{_cross_unitdir}` and `%exclude %{_cross_unitdir}` sections below.

Anything excluded should have a comment saying why it's excluded.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%exclude %{_cross_libdir}/systemd/system/sockets.target.wants/systemd-creds.socket
+%exclude %{_cross_libdir}/systemd/system/soft-reboot.target
+%exclude %{_cross_libdir}/systemd/system/ssh-access.target
+%exclude %{_cross_libdir}/systemd/system/sysinit.target.wants/systemd-tmpfiles-setup-dev-early.service
```
Why exclude `systemd-tmpfiles-setup-dev-early.service`?

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,799 @@
+%exclude %{_cross_libdir}/systemd/system/soft-reboot.target
+%exclude %{_cross_libdir}/systemd/system/ssh-access.target
+%exclude %{_cross_libdir}/systemd/system/sysinit.target.wants/systemd-tmpfiles-setup-dev-early.service
+%exclude %{_cross_libdir}/systemd/system/systemd-creds.socket
+%exclude %{_cross_libdir}/systemd/system/systemd-creds@.service
```
Can we keep these? Not sure why they're excluded.

bcressey commented:
> To forcibly reenable cgroup v1 support, SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1 must additionally be set on the kernel command line.

It may be worth a patch to remove this requirement, since it's relatively annoying to incorporate into variant definitions:
```
bool cg_is_legacy_force_enabled(void) {
    return cg_is_legacy_enabled();
}

vigh-m commented:
⬆️ Updated based on the comments above. Still pending (potentially follow up PRs):
- Validate reboots with and without `9001-core-add-separate-timeout-for-system-shutdown.patch`
- Validate the `mount rate limit` with a warning log and move it to a bootconfig drop-in using `SYSTEMD_DEFAULT_MOUNT_RATE_LIMIT_BURST`
- Patch out the need for `SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1` since it already requires `"systemd.unified_cgroup_hierarchy=0"`
- Build `cryptsetup` after ingesting the upstream changes for `openssl/ui.h`

bcressey commented on packages/release/modprobe-no-exit.conf:
```diff
@@ -0,0 +1,3 @@
+
```
```suggestion
```

bcressey commented on packages/release/release.spec:
```diff
@@ -100,6 +101,12 @@ Source1102: systemd-networkd-service-env.conf
 Source1104: aws-config.conf
 Source1105: wait-for-selinux-policy.conf
+Source1106: systemd-resolved-service-private-tmp.conf
+Source1107: issue
```
nit: I would put `issue` with "core config files"

bcressey commented on packages/release/release.spec:
```diff
@@ -175,12 +182,13 @@ Requires: (%{name}-fips if %{_cross_os}image-feature(fips))
 install -d %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}
 install -p -m 0644 %{S:11} %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}
+# Add art to the console
+install -p -m 0644 %{S:1107} %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}/issue
```
after renumbering the source, you can merge with the previous line and drop the comment:
```suggestion
install -p -m 0644 %{S:11} %{S:12} %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}
```

bcressey commented on packages/release/release.spec:
```diff
@@ -198,6 +206,12 @@ install -p -m 0644 %{S:1104} %{buildroot}%{_cross_unitdir}/service.d/00-aws-conf
+install -p -m 0644 %{S:100} %{buildroot}%{_cross_libdir}/modules-load.d/nf_conntrack.conf
+
+install -d %{buildroot}%{_cross_libdir}/systemd/journald.conf.d
+install -p -m 0644 %{S:101} %{buildroot}%{_cross_libdir}/systemd/journald.conf.d/80-journald.conf
```
The previous path is hard-coded in a `bloodhound` check (`JOURNALD_CONF_FILE`) so either that implementation needs to be smarter or we should keep the path the same:
```suggestion
install -p -m 0644 %{S:101} %{buildroot}%{_cross_libdir}/systemd/journald.conf.d/journald.conf
```

We should really be using `systemd-analyze cat-config systemd/journald.conf` there but that's more of an issue for the backlog.

vigh-m replied:
Gotcha! I was trying to avoid a situation like we saw recently with the `max-delay`

bcressey commented on packages/release/release.spec:
```diff
@@ -229,11 +243,33 @@ install -p -m 0644 %{S:1100} \
+install -p -m 0644 %{S:1105} \
+  %{buildroot}%{_cross_unitdir}/systemd-udev-trigger.service.d/00-selinux.conf
+
+install -d %{buildroot}%{_cross_unitdir}/systemd-sysusers.service.d
+install -p -m 0644 %{S:1107} \
+  %{buildroot}%{_cross_unitdir}/systemd-sysusers.service.d/50-systemd-sysusers.conf
```
`%{S:1107}` is `issue` - this would be `%{S:1109}` unless it's renumbered after `issue` moves.

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,783 @@
+License: GPL-2.0-or-later AND GPL-2.0-only AND LGPL-2.1-or-later
+URL: https://www.freedesktop.org/wiki/Software/systemd
+Source0: https://github.com/systemd/systemd/archive/v%{version}/systemd-%{version}.tar.gz
+Source1: systemd-modprobe.conf
+Source2: tmp-mount.conf
```
```suggestion
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,783 @@
+ -Dldconfig=true
+ -Dresolve=true
+ -Defi=true
+ -Dtpm=false
```
move to `-Dtpm=false` to the `tpm` area

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,783 @@
+
+%exclude %{_cross_bindir}/bootctl
+%exclude %{_cross_bindir}/run0
+%exclude %{_cross_libdir}/pam.d/systemd-run0
```
This can go down near:
```
%exclude %{_cross_libdir}/pam.d/systemd-user
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,783 @@
+%exclude %{_cross_libdir}/systemd/repart/*
+%exclude %{_cross_libdir}/systemd/system-generators/systemd-ssh-generator
+%exclude %{_cross_libdir}/systemd/system-generators/systemd-gpt-auto-generator
+
```
```suggestion
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,783 @@
+%exclude %{_cross_unitdir}/capsule.slice
+%exclude %{_cross_unitdir}/capsule@.service
+
+# Soft reboot of userspace is not implemented and supported
```
nit: phrasing
```suggestion
# Soft reboot of userspace is not implemented or supported
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,783 @@
+%exclude %{_cross_unitdir}/soft-reboot.target
+%exclude %{_cross_unitdir}/systemd-soft-reboot.service
+
+# Disable auto load of credentials via udevadm
+%exclude %{_cross_unitdir}/systemd-udev-load-credentials.service
```
Does this cause problems if we keep it?

vigh-m replied:
Not that I have experienced

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,783 @@
+%{_cross_libdir}/udev/fido_id
+%{_cross_libdir}/udev/mtd_probe
+%{_cross_libdir}/udev/scsi_id
+%exclude %{_cross_libdir}/udev/v4l_id
+%{_cross_libdir}/udev/iocost
```
nit: includes come before excludes
```suggestion
%{_cross_libdir}/udev/iocost
%exclude %{_cross_libdir}/udev/v4l_id
```

bcressey commented on packages/systemd-257/systemd-257.spec:
```diff
@@ -0,0 +1,783 @@
+%exclude %{_cross_udevrulesdir}/73-seat-late.rules
+%exclude %{_cross_udevrulesdir}/78-sound-card.rules
+%exclude %{_cross_datadir}/mime/packages/io.systemd.xml
+%exclude %{_cross_datadir}/factory/etc/vconsole.conf
```
This can move to the `%{_cross_factorydir}` excludes.

vigh-m commented:
⬆️ Pushed changes based on comments

vigh-m commented:
Additional testing notes:
- Retested sample workloads on `aws-k8s-1.33` variants with `systemd-252` to ensure that there was no regression
- Tested sample workloads on `aws-k8s-1.33-fips` variants with `systemd-257` and pods launched without issue

KCSesh commented on packages/release/release.spec:
```diff
@@ -23,7 +24,8 @@ Source95: release-systemd-networkd.conf
 Source96: release-repart-local.conf
 Source98: release-systemd-system.conf
 Source99: release-ca-certificates-tmpfiles.conf
-Source100: release-snapshotter-tmpfiles.conf
```
You have several changes where you are removing snapshotter - which we want to keep.

vigh-m replied:
Yup. This was made based on the comment [here](https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/581#discussion_r2274478980). Only the `snapshotter-tmpfiles.conf` is moved to merge with the `release-tmpfiles.conf`

KCSesh commented on packages/systemd-257/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+releases-url = "https://github.com/systemd/systemd/releases"
+
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/systemd/systemd/archive/v257.7/systemd-257.7.tar.gz"
```
Not a blocker: but 257.8 was released.

vigh-m replied:
Yeah. I saw that but was thinking of pulling that in with the next PR for cryptsetup. I think 257.8 will have the upstream openssl/ui.h fix

KCSesh commented on packages/systemd-257/9007-systemd-networkd-Conditionalize-hostnamed-timezoned-.patch:
```diff
@@ -0,0 +1,125 @@
+ int manager_request_product_uuid(Manager *m);
+-
++#else
++static inline int manager_request_product_uuid(_unused_ Manager *m){
```
Nit format on this file: 

```suggestion
+static inline int manager_request_product_uuid(_unused_ Manager *m) {
```

KCSesh commented on packages/systemd-257/9001-core-add-separate-timeout-for-system-shutdown.patch:
```diff
@@ -0,0 +1,84 @@
+         ExecOutput std_output, std_error;
+
+-        usec_t restart_usec, timeout_start_usec, timeout_stop_usec, timeout_abort_usec, device_timeout_usec;
++        usec_t restart_usec, timeout_start_usec, timeout_stop_usec, timeout_abort_usec, device_timeout_usec, timeout_shutdown_usec ;
```
nit:

```suggestion
+        usec_t restart_usec, timeout_start_usec, timeout_stop_usec, timeout_abort_usec, device_timeout_usec, timeout_shutdown_usec;
```

KCSesh commented:
For commit: ` systemd: initial build of v257`

Can we add context to  systemd-252 patch references:

> Systemd-252 patches 1001, 1002, and 1003 are no longer needed as they are included
upstream.

vigh-m commented:
⬆️ Rebased and updated based on latest comments
